File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
  let dx = [-1, 1, 0, 0];
  let dy = [0, 0, -1, 1];



  // startX = 0 , startY = 0, 
  // 움직임  n -1, m -1 에 도착하면, 
  let n = maps.length; //행
  let m = maps[0].length;  //열 

  const visited = Array.from(
    { length: n },
    () => new Array(m).fill(false)
  );

  //도착할 방법이 없음 -1  
  let q = [];
  q.push([0, 0, 1]) //y, x , dist  
  visited[0][0] = true;
  let answer = -1

  while (q.length > 0) {
    let [cy, cx, cost] = q.shift();

    if (cy === n - 1 && cx === m - 1) {
      answer = cost
    }
    for (let i = 0; i < 4; i++) {
      let nx = dx[i] + cx;
      let ny = dy[i] + cy;
      if (nx >= 0 && nx < m && ny >= 0 && ny < n && maps[ny][nx] === 1 && !visited[ny][nx]) {
        visited[ny][nx] = true;
        q.push([ny, nx, cost + 1]);
      }
    }
  }

  return answer
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    # 최단거리는 queue bfs
    q = deque()
    visited = [[False] * len(maps[0]) for _ in range(len(maps))]
    dx = [-1,1,0,0] # 이동
    dy = [0,0,-1,1] # 상하 좌우 이동이 가능해야함
    q.append([0,0,1]) # 첫칸은 한칸으로 침 

    visited[0][0] = True # 이미 청므 방문했음 
    flags = False
    
    while q:
        curx, cury, dist = q.popleft()
             
        if curx == len(maps) -1 and cury == len(maps[0]) - 1:
            flags = True
            return dist
    
        # 상하좌우 움직여야함
        for i in range(4):
          nx = dx[i] + curx
          ny = dy[i] + cury
          if nx >= 0 and nx < len(maps) and ny >= 0 and ny < len(maps[0]) and not visited[nx][ny] and maps[nx][ny] == 1:
            visited[nx][ny] = True
            q.append([nx, ny, dist +1])
            
    if not flags:
        return -1
  